tests/conftest.py

```diff
@@ -56,3 +63,271 @@ def caplog(caplog: LogCaptureFixture):
+
+        # Inspired by https://stats.stackexchange.com/a/112671/
+        def objective(x):
+            # np.exp ensures they are always positive
+            a, b = np.exp(x)
+            dist = scipy.stats.beta(a=a, b=b)
+
+            squared_error_lower = self._quantile_squared_error(dist, lower_bound, 0.025)
+            squared_error_upper = self._quantile_squared_error(dist, upper_bound, 0.975)
+
+            try:
+                return squared_error_lower + squared_error_upper
+            except FloatingPointError:
+                return np.finfo(float).max
+
+        # It is quite important to start with a reasonable guess.
+        uncertainty_interval_midpoint = (lower_bound + upper_bound) / 2
+        # TODO: Further refine these concentration values. As long as we get convergence
+        # with one of them (for all the assertions we do), we're good -- but this specific
+        # list may not get us to convergence for future assertions we add.
+        for first_guess_concentration in [10_000, 1_000, 100, 10, 1, 0.5]:
+            optimization_result = scipy.optimize.minimize(
+                objective,
+                x0=[
+                    np.log(uncertainty_interval_midpoint * first_guess_concentration),
+                    np.log((1 - uncertainty_interval_midpoint) * first_guess_concentration),
+                ],
+            )
+            # Sometimes it warns that it *may* not have found a good solution,
+            # but the solution it found is very accurate.
+            if optimization_result.success or optimization_result.fun < 1e-05:
+                break
+
+        assert optimization_result.success or optimization_result.fun < 1e-05
+
+        result = np.exp(optimization_result.x)
+        assert len(result) == 2
+        return tuple(result)
+
+    def _quantile_squared_error(
+        self, dist: scipy.stats.rv_continuous, value: float, intended_quantile: float
+    ) -> float:
+        with np.errstate(under="ignore"):
+            actual_quantile = dist.cdf(value)
+
+        if 0 < actual_quantile < 1:
+            return (
+                scipy.special.logit(actual_quantile) - scipy.special.logit(intended_quantile)
+            ) ** 2
+        else:
+            # In this case, we were so far off that the actual quantile can't even be
+            # precisely calculated.
+            # We return an arbitrarily large penalty to ensure this is never selected as the minimum.
+            return np.finfo(float).max
+
+    def save_diagnostic_output(self) -> None:
+        """
+        Save diagnostics for optional human inspection.
+        Can be useful to get more information about warnings, or to prioritize
+        areas to be more thorough in manual V&V.
+        """
+        output = pd.DataFrame(self.proportion_test_diagnostics)
+        output.to_csv(
+            Path(os.path.dirname(__file__))
+            / "v_and_v_output/proportion_test_diagnostics.csv",
```

Comment:
Do we really want these output diagnostics to be saved to the repo like this? I get that we're gitignoring but...

Reply:
I'm open to suggestions on this, but didn't immediately think of a better location. I'd prefer not to save outside the repository in some absolute path, since then we are depending on that existing.